exercise/cache/points.py

```diff
@@ -113,23 +123,36 @@ def r_augment(children):
                 #    - current submission is unofficial AND
                 #    - current best is unofficial
                 #    - current submission has better grade
-                if (
-                    ready and (
-                        entry['unofficial'] or
-                        is_better_than(submission, entry)
-                    )
-                ) or (
-                    unofficial and
-                    not entry['graded'] and # NOTE: == entry['unofficial'], but before any submissions entry['unofficial'] is False
-                    is_better_than(submission, entry)
-                ):
+                if submission.force_exercise_points:
+                    # This submission is chosen as the best submission and no
+                    # further submissions are considered.
                     entry.update({
                         'best_submission': submission.id,
                         'points': submission.grade,
-                        'passed': ready and submission.grade >= entry['points_to_pass'],
-                        'graded': ready, # != unofficial
-                        'unofficial': unofficial,
+                        'passed': submission.grade >= entry['points_to_pass'],
+                        'graded': True,
+                        'unofficial': False,
```

Comment:
What happens if the teacher tries to mark an unofficial submission as the "final" one in the manual assessment UI? Did you test that or think about it?

Reply:
I didn't think about that. Looks like it makes the UI a bit confusing: the points appear to be color coded as grey in some places, and not grey in other places. For example, submission 4 is marked as final here:
![image](https://user-images.githubusercontent.com/28843787/136018869-f1d0cadd-d968-49d1-87be-3daa2213e690.png)
I think there are 3 options to solve this:
1. Marking a submission as final changes its status from 'unofficial' to 'ready'. I think this would make sense because the label for 'unofficial' is 'no effect on grading', but that is not true when a submission marked as final: it is literally the only thing that has any effect an grading.
2. Can't mark unofficial submissions as final.
3. Display unofficial submissions, that are marked as final, consistently as official everywhere in the UI, despite being stored as unofficial submissions.

We _could_ also do the reverse of option 3: display them as unofficial everywhere in the UI. But that wouldn't really make sense, because the definitions of 'unofficial' and 'final' are conflicting, as explained in point 1.

Reply:
Option 1 makes a lot of sense. If a teacher wants to turn an unofficial submission into a graded submission, they could do it by marking the submission as final. (There are use cases for that.) The submission status can be then changed to "ready" from "unofficial".

Please test that this works well. We haven't changed unofficial submissions to graded submissions in the past. Pay attention to the student's and teacher's UI, points cache, all results table and the points API.
In addition, does it matter if the student has several normal and unofficial submissions? There could be a few normal submissions, a few unofficial submissions. If one unofficial submission is turned to a normal graded submission and there are unofficial submissions between the normal submissions, I hope that doesn't cause any issues.

Reply:
Marking a submission as final should always set the submission status to "ready". If we manually assess a submission that is stuck in the statuses "initialized" or "waiting for grading" and mark it as final, then the status should be "ready" after that.

Reply:
If a teacher manually assesses as submission and marks it as final, then is it still possible that the automatic grading overrides the points if the exercise service sends the points after the manual assessment?

Reply:
Good points. I'll implement option 1 (submissions are always set to status 'ready' when marking them as final) and test it throughly.

> If a teacher manually assesses as submission and marks it as final, then is it still possible that the automatic grading overrides the points if the exercise service sends the points after the manual assessment?

Yes, this will happen. When points are received from the grader, the status is always set to 'ready', unless it was 'unofficial'. We probably don't want that to happen, and instead the points given by the teacher should remain final.

Another thing that came to my mind: what if the teacher marks one unofficial submission as final, then decides that another unofficial submission should be graded instead, and marks that one as final? In that case, the first submission is no longer final, but it won't be unofficial either - it will remain in the 'ready' state. This doesn't affect grading, because the second submission is now final, but the original submission has still changed its state, with no way to change it back (apart from the admin UI). We can't really change it back automatically when the second submission is marked as final, because we don't store the previous state of the first submission before it was marked as final.

Reply:
**Automatic grading after manual assessment:** this is tricky since sometimes teachers might want to regrade submissions automatically after writing manual feedback. The database stores only one grade and we don't really know if it has been set automatically or manually. I would say we could keep this simple at first and ignore this, in other words, allow automatic grading to override the grade. It should be rare that people somehow accidentally auto-grade submissions after manual assessment (so it should be intentional when auto-grading really happens). Usually auto-grading has finished before manual assessment is started and the manual grade is then anyway the last one. There might also be situations when teachers want to mark one submission as final, but they want to use the auto-grading results without any manual assessment (especially if they want to convert unofficial submissions to graded submissions).

**Multiple unofficial submissions as final:** let's assume there is no problem in leaving the first unofficial submission in the ready state (which makes it official). If it doesn't break anything, then it's OK. Since unofficial submissions exist, the student has already lost the opportunity to make new graded submissions. Therefore, the teacher is trying to fix something manually by marking submissions as final and the relevant outcome is that the selected submission really defines the final grade.

Reply:
Now the submission is set to status 'ready' when marking it as final. I tested and did not find any issues caused by this in the student UI, teacher UI, points cache, all results page or the API.